# Preprocess the data
* follows same methods as Preprocessing 1; however adds 2018 individual variables to X
* Saves new X as X_alt


### Import Packages

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import sys
import os
import matplotlib.pyplot as plt
import klib
import plotly.figure_factory as ff
from IPython.core.display import display
from scipy import stats
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn import preprocessing

# pd.set_option('display.max_columns', None)
# pd.reset_option('max_rows')
#np.set_printoptions(threshold=sys.maxsize)

plt.style.use('dark_background')
plt.rcParams.update({"grid.linewidth":0.5, "grid.alpha":0.5})
sns.set(style='ticks', context='talk')

In [2]:
# define useful function to create scatterplots of target feature against desired columns
def scatterplots(df, target, columns, ncol=None, figsize=(20, 25)):
    if ncol is None:
        ncol = len(columns)
    nrow = int(np.ceil(len(columns) / ncol))
    fig, axes = plt.subplots(nrow, ncol, figsize=figsize, squeeze=False)
    fig.subplots_adjust(wspace=0.7, hspace=0.7)
    for i, col in enumerate(columns):
        # ax = axes.flatten()[i]
        # ax.scatter(x = col, y = target, data=df, alpha=0.5)
        sns.regplot(ax=axes.flatten()[i], x=col, y=target, data=df, scatter_kws={'alpha':0.5})
        # ax.set(xlabel=col, ylabel=target)
    nsubplots = nrow * ncol
    for empty in range(i+1, nsubplots):
        axes.flatten()[empty].set_visible(False)

### Load data

In [3]:
pm10_2017_2019 = r'../../data/processed/pm10_2017_2019'
pm10_year_diff = r'../../data/processed/pm10_year_diff'
pm10_year_pct_change = r'../../data/processed/pm10_year_pct_change'
pm10_year_diff_top_corr =  r'../../data/processed/pm10_year_diff_top_corr'
pm10_year_pct_change_top_corr = r'../../data/processed/pm10_year_pct_change_top_corr'

df_norm = pd.read_csv(pm10_2017_2019)
df_year_diff = pd.read_csv(pm10_year_diff)
df_year_diff_top = pd.read_csv(pm10_year_diff_top_corr)
df_year_pct_change = pd.read_csv(pm10_year_pct_change)
df_year_pct_change_top = pd.read_csv(pm10_year_pct_change_top_corr)

In [4]:
# concatenate all features
# after concentate look at features and see if need to standardize or normalize

display(df_year_pct_change)
display(df_year_diff)

AQI_2019  AQI_2017_2018_pct_change  \
0    20.073389                  0.061135   
1    12.958515                  0.059050   
2    13.125000                  0.021854   
3    11.247911                 -0.074147   
4    17.413408                 -0.138879   
..         ...                       ...   
203  10.503937                  0.168176   
204  14.431990                  0.132224   
205  15.656425                  0.259583   
206  13.354142                  0.088415   
207   9.385135                 -0.018393   

     Civilian_labor_force_2017_2018_pct_change  Employed_2017_2018_pct_change  \
0                                     0.013844                       0.020189   
1                                     0.028336                       0.033239   
2                                    -0.001903                       0.002235   
3                                     0.006385                       0.011546   
4                                     0.003142                       0.003413   
..                                         ...                            ...   
203                                  -0.009622                      -0.006847   
204                                  -0.014080                      -0.009978   
205                                  -0.002127                       0.000323   
206                                  -0.000829                      -0.001614   
207                                  -0.006981                       0.000987   

     Unemployed_2017_2018_pct_change  Unemployment_rate_2017_2018_pct_change  \
0                          -0.128942                               -0.139535   
1                          -0.093125                               -0.128205   
2                          -0.093261                               -0.093023   
3                          -0.038812                               -0.039216   
4                          -0.001463                                0.000000   
..                               ...                                     ...   
203                        -0.102491                               -0.103448   
204                        -0.111942                               -0.100000   
205                        -0.084676                               -0.103448   
206                         0.019909                                0.027778   
207                        -0.151721                               -0.134615   

     Poverty Estimate, All Ages_2017_2018_pct_change  \
0                                          -0.038496   
1                                          -0.051402   
2                                          -0.059333   
3                                           0.128447   
4                                          -0.049222   
..                                               ...   
203                                         0.040280   
204                                        -0.003911   
205                                         0.044369   
206                                        -0.048956   
207                                        -0.110532   

     90% CI LB All Ages_2017_2018_pct_change  \
0                                  -0.045462   
1                                  -0.053703   
2                                  -0.062900   
3                                   0.155791   
4                                  -0.038434   
..                                       ...   
203                                 0.012364   
204                                -0.005839   
205                                 0.040462   
206                                -0.071456   
207                                -0.140193   

     90% CI UB All Ages_2017_2018_pct_change  \
0                                  -0.032514   
1                                  -0.049561   
2                                  -0.056488   
3                                   0.105858   
4                                  -0.056579   
..   

AQI_2019  AQI_2017_2018_diff  Civilian_labor_force_2017_2018_diff  \
0    20.073389            1.008865                                 4312   
1    12.958515            0.689292                                 5003   
2    13.125000            0.292164                                 -202   
3    11.247911           -1.114232                                  130   
4    17.413408           -2.926909                                  155   
..         ...                 ...                                  ...   
203  10.503937            1.567276                                 -466   
204  14.431990            1.676852                                -6690   
205  15.656425            3.299519                                 -481   
206  13.354142            0.921309                                  -40   
207   9.385135           -0.201293                                 -276   

     Employed_2017_2018_diff  Unemployed_2017_2018_diff  \
0                       6021                      -1709   
1                       5641                       -638   
2                        227                       -429   
3                        211                        -81   
4                        159                         -4   
..                       ...                        ...   
203                     -322                       -144   
204                    -4550                      -2140   
205                       71                       -552   
206                      -75                         35   
207                       37                       -313   

     Unemployment_rate_2017_2018_diff  \
0                                -0.6   
1                                -0.5   
2                                -0.4   
3                                -0.4   
4                                 0.0   
..                                ...   
203                              -0.3   
204                              -0.4   
205                              -0.3   
206                               0.1   
207                              -0.7   

     Poverty Estimate, All Ages_2017_2018_diff  \
0                                      -4148.0   
1                                      -2312.0   
2                                      -2748.0   
3                                       3009.0   
4                                       -946.0   
..                                         ...   
203                                      282.0   
204                                     -696.0   
205                                      847.0   
206                                     -455.0   
207                                     -954.0   

     90% CI LB All Ages_2017_2018_diff  90% CI UB All Ages_2017_2018_diff  \
0                              -4526.0                            -3770.0   
1                              -2147.0                            -2477.0   
2                              -2585.0                            -2911.0   
3                               3302.0                             2716.0   
4                               -599.0                            -1293.0   
..                                 ...                                ...   
203                               73.0                              491.0   
204                             -987.0                             -405.0   
205                              673.0                             1021.0   
206                             -555.0                             -355.0   
207                            -1000.0                             -908.0   

     Poverty Percent, All Ages_2017_2018_diff  ...  \
0                                        -0.6  ...   
1                                        -0.9  ...   
2                                        -1.2  ...   
3                                         4.2  ...   
4                                        -1.0  ...   
..                                        ...  ...   

In [5]:
df = pd.merge(df_year_diff, df_year_pct_change)
display(df)


AQI_2019  AQI_2017_2018_diff  Civilian_labor_force_2017_2018_diff  \
0    20.073389            1.008865                                 4312   
1    12.958515            0.689292                                 5003   
2    13.125000            0.292164                                 -202   
3    11.247911           -1.114232                                  130   
4    17.413408           -2.926909                                  155   
..         ...                 ...                                  ...   
203  10.503937            1.567276                                 -466   
204  14.431990            1.676852                                -6690   
205  15.656425            3.299519                                 -481   
206  13.354142            0.921309                                  -40   
207   9.385135           -0.201293                                 -276   

     Employed_2017_2018_diff  Unemployed_2017_2018_diff  \
0                       6021                      -1709   
1                       5641                       -638   
2                        227                       -429   
3                        211                        -81   
4                        159                         -4   
..                       ...                        ...   
203                     -322                       -144   
204                    -4550                      -2140   
205                       71                       -552   
206                      -75                         35   
207                       37                       -313   

     Unemployment_rate_2017_2018_diff  \
0                                -0.6   
1                                -0.5   
2                                -0.4   
3                                -0.4   
4                                 0.0   
..                                ...   
203                              -0.3   
204                              -0.4   
205                              -0.3   
206                               0.1   
207                              -0.7   

     Poverty Estimate, All Ages_2017_2018_diff  \
0                                      -4148.0   
1                                      -2312.0   
2                                      -2748.0   
3                                       3009.0   
4                                       -946.0   
..                                         ...   
203                                      282.0   
204                                     -696.0   
205                                      847.0   
206                                     -455.0   
207                                     -954.0   

     90% CI LB All Ages_2017_2018_diff  90% CI UB All Ages_2017_2018_diff  \
0                              -4526.0                            -3770.0   
1                              -2147.0                            -2477.0   
2                              -2585.0                            -2911.0   
3                               3302.0                             2716.0   
4                               -599.0                            -1293.0   
..                                 ...                                ...   
203                               73.0                              491.0   
204                             -987.0                             -405.0   
205                              673.0                             1021.0   
206                             -555.0                             -355.0   
207                            -1000.0                             -908.0   

     Poverty Percent, All Ages_2017_2018_diff  ...  \
0                                        -0.6  ...   
1                                        -0.9  ...   
2                                        -1.2  ...   
3                                         4.2  ...   
4                                        -1.0  ...   
..                                        ...  ...   

In [6]:
# Add 2018 columns to dataset

df_norm = df_norm.drop(['GeoFIPS', 'State Name', 'County Name'], axis=1)

col_2018 = df_norm.columns[187:375]

df = pd.merge(df, df_norm[col_2018])


Decide to only have a df retaining features of difference and percent changes
because these column are stationarized random variables and behave similar to independent random variables
whihc will give us more suitable results for modeling

### Perform eda to decide whether we need to standardize or normalize
### Then split data before standardize and normalize


Can tell we need to standardize already by seeing some columns with varying means
especially between Civilian_labor_force_2017_2018_diff with mean 2500 and Unemployment_rate_2017_2018_diff	-.4

In [7]:
# Run Shapiro test on all features and count number of features with a p score less than .05 \
# Features with a shapiro p score less than .05 are considered not gaussian

count = 0
for feature in df.columns:
    shapiro_test = stats.shapiro(df[feature])
    if shapiro_test.pvalue < 0.05:
        count = count + 1
print(" {} features were deemed to not have a normal distribution out of {} features".format(count, len(df)))

 1244 features were deemed to not have a normal distribution out of 208 features


In [8]:
# sns.distplot(x)
shapiro_test = stats.shapiro(df['AQI_2019'])
shapiro_test.pvalue

2.866333004281074e-10

### Result
* Standardization of data should be performed because features have discrepancy in magnitudes from looking at means of some features
* Normalization of data should be performed because majority of the features are reported wht non gaussian distributions

## Split data

In [9]:
X = df.drop('AQI_2019', axis=1)
y = df['AQI_2019']

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

## Scale Data

In [11]:
# checking whether we have negative observations

for cols in df.columns.tolist()[1:]:
     data = df.loc[df[cols] > 0]
data

AQI_2019  AQI_2017_2018_diff  Civilian_labor_force_2017_2018_diff  \
0    20.073389            1.008865                                 4312   
1    12.958515            0.689292                                 5003   
2    13.125000            0.292164                                 -202   
3    11.247911           -1.114232                                  130   
4    17.413408           -2.926909                                  155   
..         ...                 ...                                  ...   
203  10.503937            1.567276                                 -466   
204  14.431990            1.676852                                -6690   
205  15.656425            3.299519                                 -481   
206  13.354142            0.921309                                  -40   
207   9.385135           -0.201293                                 -276   

     Employed_2017_2018_diff  Unemployed_2017_2018_diff  \
0                       6021                      -1709   
1                       5641                       -638   
2                        227                       -429   
3                        211                        -81   
4                        159                         -4   
..                       ...                        ...   
203                     -322                       -144   
204                    -4550                      -2140   
205                       71                       -552   
206                      -75                         35   
207                       37                       -313   

     Unemployment_rate_2017_2018_diff  \
0                                -0.6   
1                                -0.5   
2                                -0.4   
3                                -0.4   
4                                 0.0   
..                                ...   
203                              -0.3   
204                              -0.4   
205                              -0.3   
206                               0.1   
207                              -0.7   

     Poverty Estimate, All Ages_2017_2018_diff  \
0                                      -4148.0   
1                                      -2312.0   
2                                      -2748.0   
3                                       3009.0   
4                                       -946.0   
..                                         ...   
203                                      282.0   
204                                     -696.0   
205                                      847.0   
206                                     -455.0   
207                                     -954.0   

     90% CI LB All Ages_2017_2018_diff  90% CI UB All Ages_2017_2018_diff  \
0                              -4526.0                            -3770.0   
1                              -2147.0                            -2477.0   
2                              -2585.0                            -2911.0   
3                               3302.0                             2716.0   
4                               -599.0                            -1293.0   
..                                 ...                                ...   
203                               73.0                              491.0   
204                             -987.0                             -405.0   
205                              673.0                             1021.0   
206                             -555.0                             -355.0   
207                            -1000.0                             -908.0   

     Poverty Percent, All Ages_2017_2018_diff  ...  HWAC_MALE_ratio_2018  \
0                                        -0.6  ...              1.881780   
1                                        -0.9  ...              2.302427   
2                                        -1.2  ...              1.520503   
3                                         4.2  ...              2.368571 

In [12]:
# because we negative observation we cannot perform logarithmic scale and will instead perform standardization
scaler = preprocessing.StandardScaler().fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_train_scaled

array([[ 2.70551905,  0.2327869 ,  0.16039692, ...,  0.73126611,
         0.33814272,  0.13515283],
       [-0.53530159, -0.08904047,  0.2148679 , ...,  1.66518729,
         1.19793835,  1.24820206],
       [ 0.77647735, -0.25122218, -0.38406301, ..., -0.65992905,
        -0.66346052, -0.64236676],
       ...,
       [-1.3393787 , -0.37625047, -0.53810594, ..., -0.39892099,
        -0.59093489, -0.57175185],
       [ 1.17720255, -0.26859427, -0.45315121, ..., -0.65292137,
        -0.65261229, -0.69127578],
       [-0.55905886,  0.21034268,  0.10105353, ..., -0.41542781,
        -0.52562127, -0.51581242]])

In [13]:
# scale test data using the scaler fitted from the training set

X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[ 1.17998294, -0.04909735,  0.08943472, ..., -0.49835965,
        -0.59863848, -0.46814034],
       [-0.51420716, -0.15561232, -0.27711999, ...,  0.02858419,
        -0.02457733, -0.01189596],
       [-0.43934205,  0.09875557,  0.40214312, ..., -0.62340591,
        -0.89072557, -0.85107061],
       ...,
       [ 0.03883044, -0.31677961, -0.43553559, ...,  0.34577355,
         0.65650346,  0.47933231],
       [-0.33949094,  0.96304845,  1.14674643, ..., -0.10691837,
        -0.21937037, -0.11472616],
       [ 0.05234839, -0.10793419, -0.26974892, ..., -0.03763552,
        -0.09835352, -0.08431688]])

In [14]:
# retain X_train_scale as a dataframe
X_train_scaled = pd.DataFrame(X_train_scaled, index=X_train.index, columns=X_train.columns)
display(X_train_scaled)

# retain X_test_scale as a dataframe
X_test_scaled = pd.DataFrame(X_test_scaled, index=X_test.index, columns=X_test.columns)
display(X_test_scaled)

AQI_2017_2018_diff  Civilian_labor_force_2017_2018_diff  \
29             2.705519                             0.232787   
19            -0.535302                            -0.089040   
55             0.776477                            -0.251222   
93             0.021271                            -0.363063   
181            0.105938                             0.161016   
..                  ...                                  ...   
106           -0.497668                             0.973573   
14             0.352561                            -0.026907   
92            -1.339379                            -0.376250   
179            1.177203                            -0.268594   
102           -0.559059                             0.210343   

     Employed_2017_2018_diff  Unemployed_2017_2018_diff  \
29                  0.160397                   0.184132   
19                  0.214868                  -0.808003   
55                 -0.384063                   0.364611   
93                 -0.509996                   0.406645   
181                 0.018098                   0.375286   
..                       ...                        ...   
106                 0.912621                   0.124417   
14                  0.191880                  -0.583155   
92                 -0.538106                   0.447011   
179                -0.453151                   0.503390   
102                 0.101054                   0.283545   

     Unemployment_rate_2017_2018_diff  \
29                          -0.708151   
19                          -1.896018   
55                          -0.114218   
93                           0.182749   
181                          0.776682   
..                                ...   
106                          0.479716   
14                          -0.411185   
92                           0.479716   
179                          0.776682   
102                          0.479716   

     Poverty Estimate, All Ages_2017_2018_diff  \
29                                   -0.032929   
19                                   -0.366625   
55                                    0.171585   
93                                    0.164637   
181                                  -0.053591   
..                                         ...   
106                                  -0.753437   
14                                   -1.187242   
92                                    0.511314   
179                                   0.239330   
102                                   0.624771   

     90% CI LB All Ages_2017_2018_diff  90% CI UB All Ages_2017_2018_diff  \
29                           -0.052244                          -0.012659   
19                           -0.255945                          -0.478881   
55                            0.192251                           0.148936   
93                            0.192520                           0.134550   
181                          -0.097600                          -0.007554   
..                                 ...                                ...   
106                          -0.800489                          -0.699322   
14                           -1.037827                          -1.333915   
92                            0.493196                           0.526518   
179                           0.237250                           0.239805   
102                           0.600995                           0.645045   

     Poverty Percent, All Ages_2017_2018_diff  \
29                                  -0.358545   
19                                  -0.358545   
55                                   0.144708   
93                                   0.144708   
181                                 -0.358545   
..                                        ...   
106                                 -0.861798   
14                                  -0.789905   
92                                   1.726361   
179               

AQI_2017_2018_diff  Civilian_labor_force_2017_2018_diff  \
161            1.179983                            -0.049097   
15            -0.514207                            -0.155612   
73            -0.439342                             0.098756   
96             0.772088                            -0.380181   
166           -0.232902                            -0.437497   
9              1.005701                             0.734548   
100           -0.677657                            -0.381323   
135            0.119467                            -0.633535   
18             2.321711                            -0.424943   
148           -0.946933                            -0.373714   
171           -0.249317                            -0.180339   
30            -0.367080                             2.115946   
155            0.286757                            -0.429888   
180           -0.180555                             0.981435   
125            0.440708                             1.015292   
197           -0.703990                            -0.244882   
164           -1.017940                            -0.438511   
190            1.440829                            -0.171970   
84            -0.541430                            -0.145595   
75            -0.458308                            -1.168646   
124            1.497074                             3.385757   
170            0.241661                             0.059193   
104           -0.573723                            -0.132154   
101           -1.138832                            -0.357991   
69            -0.690100                            -0.058734   
25             0.827730                            -0.331996   
95             0.400625                             0.182192   
16             0.541304                            -0.082700   
141           -0.359028                             1.000329   
185           -0.344915                            -0.160304   
154            0.336276                            -0.257309   
68            -0.684595                            -0.365092   
66            -0.941964                            -0.362048   
120           -1.293715                            -0.292941   
147            0.224571                            -0.161699   
98             0.807830                            -0.401104   
138           -0.379972                             0.120693   
167           -0.398301                            -0.513071   
45            -0.701664                            -0.397300   
113            1.568841                            -0.400090   
65            -3.733254                             0.069971   
178           -2.064845                            -0.143186   
86             0.512360                            -0.302197   
203            0.700078                            -0.380435   
67            -0.553167                             0.274632   
82            -0.541137                            -0.109709   
205            1.602388                            -0.382337   
194           -0.168360                             0.743298   
38             0.261369                             0.305445   
24             0.038830                            -0.316780   
60            -0.339491                             0.963048   
195            0.052348                            -0.107934   

     Employed_2017_2018_diff  Unemployed_2017_2018_diff  \
161                 0.089435                  -0.367981   
15                 -0.277120                   0.330583   
73                  0.402143                  -0.814008   
96                 -0.551224                   0.471697   
166                -0.518741                   0.234172   
9                   0.493220                   0.615480   
100                -0.557096                   0.484374   
135                -0.372569                  -0.671893   
18                 -0.535482                   0.311902   
148                -0.

### save X and y and scaled data


In [18]:
X_path = r'../../data/train_test/X_alt'
X.to_csv(X_path, index=False)

X_train_scaled_path = r'../../data/train_test/X_train_scaled_alt'
X_train_scaled.to_csv(X_train_scaled_path, index=False)

X_test_scaled_path = r'../../data/train_test/X_test_scaled_alt'
X_test_scaled.to_csv(X_test_scaled_path, index=False)

y_path = r'../../data/train_test/y_alt'
y.to_csv(y_path, index=False)

y_train_path = r'../../data/train_test/y_train_alt'
y_train.to_csv(y_train_path, index=False)

y_test_path = r'../../data/train_test/y_test_alt'
y_test.to_csv(y_test_path, index=False)

df_path = r'../../data/processed/df_standardized.csv'
df.to_csv(df_path, index=False)

In [16]:
# param_grid = {'alpha':np.arange(0.1,2, step=0.1)}
# lasso = Lasso()
# lasso_grid = GridSearchCV(lasso, param_grid, cv=5)
# lasso_grid.fit(X_train, y_train)
# print('r2 / variance: ', lasso_grid.best_score_, 'best params', lasso_grid.best_params_)
# print('best estimator', lasso_grid.best_estimator_)

In [17]:
lasso = Lasso(alpha=1.0)
lasso.fit(X_train_scaled, y_train)
lasso.score(X_train_scaled, y_train)

0.33139496642885813

## results
When choosing LASSO and without 2018 individual variables, the training score yielded was .71.
However when including the 2018 individual variables, the training score jumped to .78.
With scaled train and test data it appears that with adding 2018 individual columns we go from .32 to .33;
which is only a small improvement for large number of columns added. Which means we will not add 2018 individual columns to
our X.
Note cross validation was not used to output these scores